pkg/scheduler/internal/cache/cache.go

```diff
@@ -477,22 +512,13 @@ func (cache *schedulerCache) AddPod(pod *v1.Pod) error {
 		if currState.pod.Spec.NodeName != pod.Spec.NodeName {
 			// The pod was added to a different node than it was assumed to.
 			klog.InfoS("Pod was added to a different node than it was assumed", "pod", klog.KObj(pod), "assumedNode", klog.KRef("", pod.Spec.NodeName), "currentNode", klog.KRef("", currState.pod.Spec.NodeName))
-			// Clean this up.
-			if err = cache.removePod(currState.pod); err != nil {
-				klog.ErrorS(err, "Error occurred while removing pod")
-			}
-			cache.addPod(pod)
 		}
-		delete(cache.assumedPods, key)
-		cache.podStates[key].deadline = nil
-		cache.podStates[key].pod = pod
+		if err = cache.updatePod(currState.pod, pod, key); err != nil {
```

Comment:
I'm not convinced that this yields the same result. I think you shouldn't try to use `updatePod` here.

Reply:
> 

Thanks a lot for pointing out. Let's dig into the details.

Firstly, the `addPodToNode` and `removePodFromNode`  functions will do these things:

- `addPodToNode`:  add pod to node and modify the **nodeInfoListItem**.
- `removePodFromNode`: delete pod from node and modify the **nodeInfoListItem**.

Secondly, let's analyze two cases:

| Cases                                                        | Before Editing                                               | After Editing                                                | Same Result ?                                                |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **Case 1:**<br> `currState.pod.Spec.NodeName != pod.Spec.NodeName` | **Step 1:** delete the oldPod from **node**  <br>**Step 2:** add the newPod to **node** <br>**Step 3:** delete the key from `assumedPods` and reset the `podStates` | One `updatePod` means:<br>**Step 1:** delete the oldPod from **cache** (not only **node** but also  `assumedPods`  and `podStates`) <br/>**Step 2:** add the newPod to cache (not only **node** but also `podStates`) | **Yes**<br>- `assumedPods`: same (both delete the key)<br>- `podStates`:  same (both set the value of `podStates[key]` to newPod)<br>- `nodeInfoListItem`: same (both modify the `nodeInfoListItem`) |
| **Case 2:**<br> `currState.pod.Spec.NodeName == pod.Spec.NodeName` | **Step 1:** delete the key from `assumedPods` and reset the `podStates` | One `updatePod` means:<br/>**Step 1:** delete the oldPod from **cache** (not only **node** but also  `assumedPods`  and `podStates`) <br/>**Step 2:** add the newPod to cache (not only **node** but also `podStates`) | **Not completely**<br>- `assumedPods`: same (both delete the key)<br/>- `podStates`:  same (both set the value of  `podStates[key]` to newPod)<br/>- `nodeInfoListItem`: **different** (After editing we will modify the `nodeInfoListItem` in `removePod` and `addPod`, while before editing we won't) |

The question is:

Should we always update the `nodeInfoListItem` when we handling add event of an assumed pods?

> Before editing, we won't update the `nodeInfoListItem` when `currState.pod.Spec.NodeName == pod.Spec.NodeName` .

- If the answer is YES, there is no problem after editing.
- Otherwise, we shouldn't use `updatePod` here.

Reply:
I don't think a generic update() rewards us significant benefits. In terms of case 2, which is the most common happy path, with current changes, we remove the pod and re-add the pod which imposes overhead unnecessarily.